File: backend/internal/pkg/crm/lead_status_transition.go
package crm

// Lead status transitions per phase-2-crm-ai / QA matrix (PATCH only; converted via /convert).
var leadStatusTransitions = map[string]map[string]bool{
	"new":         {"contacted": true},
	"contacted":   {"qualified": true, "unqualified": true},
	"qualified":   {"unqualified": true},
	"unqualified": {},
	"converted":   {},
}

// CanTransitionLeadStatus reports whether PATCH may change status from → to.
// Setting converted is only allowed through POST /leads/:id/convert.
func CanTransitionLeadStatus(from, to string) bool {
	if from == to {
		return true
	}
	if to == "converted" {
		return false
	}
	next, ok := leadStatusTransitions[from]
	if !ok {
		return false
	}
	return next[to]
}

// CanConvertLead reports whether the lead may be converted from its current status.
func CanConvertLead(status string) bool {
	return status == "qualified"
}

File: apps/web/utils/lead-status-transition.ts
import type { LeadStatus } from '~/types/lead'

/** Mirrors backend/internal/pkg/crm/lead_status_transition.go */
const NEXT_STATUS: Partial<Record<LeadStatus, LeadStatus[]>> = {
  new: ['contacted'],
  contacted: ['qualified', 'unqualified'],
  qualified: ['unqualified'],
  unqualified: [],
  converted: [],
}

export function canTransitionLeadStatus(from: LeadStatus, to: LeadStatus): boolean {
  if (from === to) return true
  if (to === 'converted') return false
  return NEXT_STATUS[from]?.includes(to) ?? false
}

export function allowedLeadStatusTargets(from: LeadStatus): LeadStatus[] {
  const next = NEXT_STATUS[from] ?? []
  return [from, ...next.filter((s) => s !== from)]
}

export function canConvertLead(status: LeadStatus): boolean {
  return status === 'qualified'
}
